---
layout: post
title: ETE 3 Installation Guide
author: denjixx
date: 2023-09-02 23:56:50 +0800
---

This is a guide on how to install `ete3` based upon experience. Consider this as a viable installation guide, however, it may not reflect real-world installation problems, as written in the ETE Cookbook: ***"any problem derived from the compilation of external software is out of the scope of the ETE toolkit."*** At the same time, I am writing this with ETE version 3.1.3 installed in Ubuntu 22.04 WSL2, if you are now at `ete4` (or want the functionality of `ete4`), consider checking the etetoolkit/ete repository for instructions. This pertains more with troubleshooting **MUSCLE** and **SLR**.



## Installing Miniconda3
After you `sudo apt update`, `sudo apt upgrade` (&& doesn't work for me, I am a noob), and `sudo apt install build-essential` (this is optional, I just added it), and assuming you have `wget` installed, proceed to download Miniconda3 using `wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O Miniconda3-latest-Linux-x86_64.sh`, then install using `bash Miniconda3-latest-Linux-x86_64.sh -p ~/eteconda/` to create directory `/home/{username}/eteconda` (NB: This was taken from [etetoolkit/course](https://github.com/etetoolkit/course)). Just follow the prompts on the Miniconda3 installation, they are user-friendly.

### Creating and activating a `conda` environment
After restarting the terminal, follow the instruction from the [website](http://etetoolkit.org/download/), then proceed with creating a conda environment using `conda create -n ete3 python=3`. This will create a conda environment named `ete3` running in `python=3`  (missing the `python=3` code may result in the whole thing not working, please follow religiously). Then, to activate, simply `conda activate ete3`. 

## Installation Proper
When the `ete3` conda environment is finally activated, proceed with installing `ete3` using `conda install -c etetoolkit ete3 ete_toolchain`. This will install etetoolkit packages from conda, so, just follow everything that is prompted. After a few moments, conda may prompt you to install other tools using a recommended script. Just don't worry about it for a while and try executing `ete3 build check`.

## THE HARD PART
Now comes the hard part, remember when I said this is a case-to-case basis. This is where it got messy. As for me, this thing happened after installation:

```
(ete3) rene@LAPTOP-UQTO8KCL:~$ ete3 build check
Current Toolchain path: /home/rene/eteconda/envs/ete3/bin/ete3_apps/bin
Current Toolchain version: unknown
       clustalo: OK - 1.2.4
         codeml: OK - AAML in paml version 4.8a, July 2014
      dialigntx: OK - This is DIALIGN-TX Version 1.0.2 - A Multiple Sequence alignment program.
       fasttree: OK - Usage for FastTree version 2.1.8 Double precision (No SSE3), OpenMP (1 threads):
         iqtree: OK - IQ-TREE multicore version 1.5.5 for Linux 64-bit built Jun  2 2017
         kalign: OK - Kalign version 2.04, Copyright (C) 2004, 2005, 2006 Timo Lassmann
          mafft: OK - MAFFT v6.861b (2011/09/24)
         muscle: MISSING
          phyml: OK - . This is PhyML version 20160115.
     pmodeltest: OK - pmodeltest.py v1.4
          raxml: OK - This is RAxML version 8.2.11 released by Alexandros Stamatakis on June 2017.
 raxml-pthreads: OK - This is RAxML version 8.2.11 released by Alexandros Stamatakis on June 2017.
         readal: OK - readAl v1.4.rev6 build[2012-02-02]
            slr: MISSING
         statal: OK - statAl v1.4.rev6 build[2012-02-02]
        tcoffee: OK - PROGRAM: T-COFFEE Version_11.00.8cbe486 (2014-08-12 22:05:29 - Revision 8cbe486 - Build 477)
         trimal: OK - trimAl v1.4.rev6 build[2012-02-02]

WARNING: 2 external tools seem to be missing or unfunctional
Install using conda (recomended):
 conda install -c etetoolkit ete3_external_apps
or manually compile by running:
 ete3 upgrade-external-tools

```
The only way to get out of these problems is to manually installing the missing external apps. I have a closed issue from [etetoolkit/ete](https://github.com/etetoolkit/ete/issues/657) explaining this. I'd like to share it here:


**FOR MUSCLE**

First, `sudo apt install muscle` to install *MUSCLE v3.8.1551 by Robert C. Edgar*, independent of ete3. Then, create a symbolic link by using `ln -s $(which muscle) [Current Toolchain path]` (use `ete3 build check` to know here the apps are located.) Detailed explanation is here: https://groups.google.com/g/etetoolkit/c/mlqJHJcu87E

**FOR SLR**

There is no workaround but to manually install *SLR "Sitewise Likelihood Ratio" selection detection program. Version 1.3*.  Detailed explanation and instruction is here: https://groups.google.com/g/etetoolkit/c/1rWr4GV3zO0

1. First, download SLR binary using either `wget` or `curl`. For me, I used `wget -O slr https://www.ebi.ac.uk/goldman-srv/SLR/download/v1.3/slr_linux.tgz`. 
2. After which, uncompress binary using `tar xzvf slr`.
3. Then, copy binary to ete3 using `cp slr/bin/Slr_shared [Current Toolchain Path]/Slr`.
4.  Lastly, install `libatlas3-base` by using `sudo apt install libatlas3-base`.

Recheck the build using `ete3 build update` and the final looks like this:
```
(ete3) rene@LAPTOP-UQTO8KCL:~$ ete3 build check
Current Toolchain path: /home/rene/eteconda/envs/ete3/bin/ete3_apps/bin
Current Toolchain version: unknown
       clustalo: OK - 1.2.4
         codeml: OK - AAML in paml version 4.8a, July 2014
      dialigntx: OK - This is DIALIGN-TX Version 1.0.2 - A Multiple Sequence alignment program.
       fasttree: OK - Usage for FastTree version 2.1.8 Double precision (No SSE3), OpenMP (1 threads):
         iqtree: OK - IQ-TREE multicore version 1.5.5 for Linux 64-bit built Jun  2 2017
         kalign: OK - Kalign version 2.04, Copyright (C) 2004, 2005, 2006 Timo Lassmann
          mafft: OK - MAFFT v6.861b (2011/09/24)
         muscle: OK - MUSCLE v3.8.1551 by Robert C. Edgar
          phyml: OK - . This is PhyML version 20160115.
     pmodeltest: OK - pmodeltest.py v1.4
          raxml: OK - This is RAxML version 8.2.11 released by Alexandros Stamatakis on June 2017.
 raxml-pthreads: OK - This is RAxML version 8.2.11 released by Alexandros Stamatakis on June 2017.
         readal: OK - readAl v1.4.rev6 build[2012-02-02]
            slr: OK - # SLR "Sitewise Likelihood Ratio" selection detection program. Version 1.3
         statal: OK - statAl v1.4.rev6 build[2012-02-02]
        tcoffee: OK - PROGRAM: T-COFFEE Version_11.00.8cbe486 (2014-08-12 22:05:29 - Revision 8cbe486 - Build 477)
         trimal: OK - trimAl v1.4.rev6 build[2012-02-02]
```

## Installing `jupyter-notebook`
`ete3` feels better with `jupyter-notebook` so installing a Jupyter Notebook in WSL should be worthwhile, except that WSL is a CLI and the prospects of having chrome running via WSL is such a ***pain in the bash***. [Here](https://harshityadav95.medium.com/jupyter-notebook-in-windows-subsystem-for-linux-wsl-8b46fdf0a536) is a way to install Jupyter Notebook in WSL2 and use the Windows-installed Chrome to access your `.ipynb` files:

1. Update Ubuntu using `sudo apt update` and `sudo apt upgrade`. Then, when all are good, install `sudo apt install python3 python3-pip ipython3`. To check what python version was installed, simply use `python3 --version`. To get rid of the extra **3** after python or pip, you can install `sudo apt install python-is-python3`. This should allow you to use `pip` or `python` and get a response.
2. Next, install Jupyter Notebook using `pip install jupyter`.
3. Afterwards, open `~/.bashrc` using either `nano` or `vim` and add `alias jupyter-notebook="~/.local/bin/jupyter-notebook --no-browser"` to the end of the bash config file.
4. Run Jupyter Notebook using `jupyter-notebook`, copy the URL and paste it in your browser to access the interface and notebooks.
5. **VOILA!** You now have all the necessary things to be able to do phylogenetic analysis!

## NEXT STEPS
Consider reading the whole [ETE Cookbook](http://etetoolkit.org/cookbook/) to be familiar with all the commands in `ete3`, including tree building via `ete-build` and testing evolutionary hypothesis via `ete-evol`. Also, ETE has a handbook for the [Python API](http://etetoolkit.org/docs/latest/tutorial/index.html), [ETE Reference Guide](http://etetoolkit.org/docs/latest/reference/index.html), and [Phylogenomic Tools](http://etetoolkit.org/documentation/tools/). **LASTLY, CONSIDER HAVING A LIFE. JUST HAVE ONE.**

## CITATION
All external applications used are cited after each workflow. Cite them in your references!!!

**ETE3 is made possible by:**
```
Huerta-Cepas J, Serra F and Bork P. ETE 3: Reconstruction, analysis and
       visualization of phylogenomic data. Mol Biol Evol (2016)
       doi:10.1093/molbev/msw046
```
